src/linker/Linker/MarkingHelpers.cs

```diff
@@ -11,12 +11,34 @@ public MarkingHelpers (LinkContext context)
 			_context = context;
 		}
 
-		public void MarkExportedType (ExportedType type, ModuleDefinition module, in DependencyInfo reason)
+		public void MarkExportedType (ExportedType exportedType, ModuleDefinition module, in DependencyInfo reason)
 		{
-			if (!_context.Annotations.MarkProcessed (type, reason))
+			if (!_context.Annotations.MarkProcessed (exportedType, reason))
 				return;
-			if (_context.KeepTypeForwarderOnlyAssemblies)
-				_context.Annotations.Mark (module, new DependencyInfo (DependencyKind.ModuleOfExportedType, type));
+
+			AssemblyDefinition facadeAssembly = module.Assembly;
+			TypeDefinition resolvedType = exportedType.Resolve ();
+			AssemblyAction facadeAction = _context.Annotations.GetAction (facadeAssembly);
+			if (facadeAssembly != null && resolvedType != null) {
+				while ((facadeAssembly != resolvedType.Module.Assembly) &&
+						facadeAction == AssemblyAction.Copy || facadeAction == AssemblyAction.CopyUsed) {
+					_context.Annotations.Mark (module, new DependencyInfo (DependencyKind.ModuleOfExportedType, exportedType));
+					foreach (var assemblyReference in module.AssemblyReferences) {
```

Comment:
I think this needs to resolve the pointed-to thing and mark it if it's a forwarder (not just mark the module). I would do it by extracting this into a helper that resolves a forwarder through only one level of indirection, to the next forwarder.